src/dune_lang/string_with_vars.ml

```diff
@@ -295,14 +310,27 @@ module Make_expander (A : Applicative) : Expander with type 'a app := 'a A.t = s
             | Pform (source, p) ->
               let+ v = f ~source p in
               Result.map v ~f:(fun v ->
-                if t.quoted
-                then [ Value.L.concat ~dir v ]
-                else (
-                  let vs = Mode.value Many v ~source in
-                  List.map ~f:(Value.to_string ~dir) vs))))
+                Value.Deferred_concat.concat_values v ~sep:inner_sep)))
       in
       Result.map (Result.List.all chunks) ~f:(fun chunks ->
-        Mode.string mode (String.concat (List.concat chunks) ~sep:""))
+        `Deferred_concat (Value.Deferred_concat.concat chunks ~sep:None))
+  ;;
+
+  let expand_result
```

Comment:
I wonder, could maybe `expand_result` be replaced by `expand_result_deferred_concat`? To avoid introducing the need to distinguish between `Value.t` and `Value.Deferred_concat.t` (so `Value.t` becomes the deferred version and is then forced). I would prefer not adding additional concept of deferred values, so if values are deferred by default and always have to be forced that could keep the mental model simpler (since `expand_result_deferred_concat` called under the hood anyway).

Reply:
Interesting idea. Making values deferred by default might tidy up some of the messiness introduced in this change.